sessionmanager.md

```diff
@@ -0,0 +1,157 @@
+like to allow for **many** smaller streams of messages within a single
+queue or subscription.
+
+## Use case
+
+A simple use case would be having a set of documents. Users can "edit"
+documents, and each edit is stored as a message within Service Bus.
+
+```javascript
+{
+  sessionId: "the document name"
+  body: {
+    mode: "edit",
+    content: "content of the edit"
+  }
+}
+```
+
+Multiple edits or transformations could be sent for a document, at any time.
+
+To handle this the user would need to write code similar to this:
+
+```javascript
+while (true) {
+  // some sort of check to make sure we're not processing
+  // too many sessions at a time. The odds of having more documents
+  // than available processors is _high_.
+  acquireSemaphore();
+
+  // by not specifying a session id we will get the next unlocked session from service bus.
+  const sessionReceiver: SessionReceiver = serviceBusClient.getSessionReceiver(
+    "queue",
+    "peekLock"
+  );
+  startSessionLockAutoRenewal(sessionReceiver);
+
+  sessionReceiver.subscribe(messageHandlers);
+
+  // ...after some predefined period of time
+  sessionReceiver.close();
+
+  stopSessionLockRenewal(sessionReceiver);
+  releaseSemaphore();
+}
+```
+
+## Proposal
+
+In the same vein as what we provide in `Receiver.subscribe()` we can provide a
+`SessionManager` that would support the use case where the user would like
+to read messages, in parallel, from a number of sessions while controlling
+the amount of parallelism.
+
+This would simplify session usage to the point where it becomes more accessible,
+allowing more customers to use this powerful feature.
+
+The proposed interface for this change:
+
+```javascript
+class ServiceBusClient {
+  // a session manager is conceptually a receiver, but does not have to
+  // obey the same interface.
+  createSessionManager("queue", "peekLock", sessionManagerOptions): SessionManager;
+  createSessionManager("topic", "subscription", "peekLock", sessionManagerOptions): SessionManager;
+}
+
+interface SessionManagerOptions {
+  // amount of time we will wait for messages to arrive in a session before
+  // it is `close()`d
+  idleTimeoutMs: number;
+
+  // the maximum number of receivers that can be active at any
+  // time. Allows the customer to avoid getting steamrolled by too many
+  // sessions.
+  maxConcurrentReceivers: number;
+
+  // new feature (and probably pushed down to SessionReceiver as well) to enable auto-renewal
+  // of the session lock.
+  maxSessionAutoRenewLockDurationInMs?: number;
+
+  // quite a few of the options available in MessageHandler (used for normal Receiver subscribe)
+  // would also apply here.
+  autoComplete?: boolean;
+}
+
+interface SessionMessageHandler {
+  processMessage(message: ReceivedMessage|ReceivedMessageWithLock);
+
+  // tentatively suggesting proposal #2 for error handling below.
+  processError(err: Error, context: { sessionId: string });
```

Comment:
Can we have the setState() and getState() methods pulled into the context?
When using the SessionManager, I would see the need to set/get state as part of the message processing code path.

Reply:
I think if we wanted to maintain parity between what we provide in processError that makes sense. I've always favored the context model.... :)

Reply:
I've added in a "context" object (which is really just a subset of SessionReceiver). 

I've left out the full context object from processError. It seems like a lot of cases (maybe all?) where processError would be called would also be leaving the session in an unusable state (ie, connection/receiver being closed, etc..)